lib/mix/lib/mix/config.ex

```diff
@@ -155,10 +163,18 @@ defmodule Mix.Config do
   `file` is the path to the configuration file to be read. If that file doesn't
   exist or if there's an error loading it, a `Mix.Config.LoadError` exception
   will be raised.
+
+  `loaded_paths` is a list of configuration files that have been previously
+  read. If `file` exists in `loaded_paths`, a `Mix.Config.LoadError` exception
+  will be raised.
   """
-  def read!(file) do
+  def read!(file, loaded_paths \\ []) do
     try do
-      {config, binding} = Code.eval_file(file)
+      if file in loaded_paths do
+        raise ArgumentError, message: "Recursive load of #{file} detected."
```

Comment:
We use lowercase first letter and no final period for raised messages (except `Mix.raise`) :)